### PR TITLE
Implement metric collection (resolves #1)

### LINKED ISSUE #1: Add prometheus /metrics support
The idea would be to be able to peek into what's going to give insights into the various measurables, like number of pods matched, filtered and actioned on, the number of machines taken up and down, etc.

### PR DESCRIPTION
**Pull request by kz**
This PR adds metric collection of events which users may consider useful to be monitored. Like cloud drivers, there can be multiple metric collectors subscribing to an AbstractCollector interface describing how metrics should be stored.

Two metric collectors have been implemented:
- StdoutCollector, a "no-op" collector which simply prints that an event has occurred to stdout
- PrometheusCollector, which saves metrics to the default registry

The metric collectors collect the following events:

| Metric | Labels (Metadata) | Description | Justification |
| - | - | - | - |
| seal_pod_kills_total | status, namespace, name | Number of pods killed (including failures) | Example alerts include number of pod kills failing in the past five minutes or the ratio between successful and failed pod kills in a time range. If a pod cannot be killed, it could be an an unresponsive state or there may be a problem with keeping Kubernetes synchronised to the node. |
| seal_nodes_stopped_total | status, uid, name| Number of nodes killed (including failures) | Example alerts include number of node stops failing in the past five minutes or the ratio between successful and failed node stops. If a node cannot be stopped, it could be in an unresponsive state. |
| seal_execute_failed_total | uid, name | Tracks failure of command executions | Commands executions failing is a general case of errors which should be brought to the user's attention. |
| seal_empty_filter_total | N/A | Cases where filtering returns an empty result | If the user's policy is designed to model common levels of failures, then having no nodes/pods after filtering could mean that insufficient resources have been provisioned for the system to withstand failure. |
| seal_probability_filter_not_passed_total | N/A | Cases where the probability filter decides to skip all nodes | Useful to track long-term in order to ensure that probability distribution is as expected. |
| seal_empty_match_total | source (either `nodes` or `pods`) | Cases where matching returns an empty result | See `seal_empty_filter_total` |

Metrics are only collected in Autonomous mode as most metrics do not apply without a policy file. For metrics which do apply without a policy file (pod/node kill failures and command execution failures), issues are immediately shown in the terminal in interactive mode.

Users can choose either the `--stdout-collector` (**StdoutCollector is the default collector**) or `--prometheus-collector` flags. If the latter is used, the `--prometheus-host` and `--prometheus-port` arguments should be specified. In Autonomous mode, the Prometheus collector will start a web server with metrics display under the `/metrics` path.

# Implementation
- The AbstractCollector class defines an interface for all the metrics which can be added. This class and the classes which extend it reside in the `metriccollectors` directory
- Tests reside in `tests/metriccollectors`
- `cli/__main__.py` has been modified to support the command line arguments
- The rest of the code changes add function calls to the abstract collector

Tests are all passing.

# Prometheus + Grafana + AlertManager examples

The metrics which are stored by the PrometheusCollector can be scraped by Prometheus for monitoring and alerting.

Here is an example of a Grafana panel with the collected metrics:
Grafana screenshot

An example AlertManager configuration could be as follows:

`/etc/seal.alerts.yml`:
```yaml
- name: PowerfulSeal Alerts
  rules:
  - alert: pod_kills_failing
    expr: rate(seal_pod_kills_total{status="failure"}[5m]) > 0
    for: 5m
    annotations:
      summary: Pod kills failing for {{ $labels.namespace }}/{{ $labels.name }}
      description: >
        PowerfulSeal has been unable to kill {{ $labels.namespace }}/{{ $labels.name }}
        for a while now.
    labels:
     type: seal
  - alert: node_kills_failing
    expr: rate(seal_node_stopped_total{status="failed"}[5m]) > 0
    for: 5m
    annotations:
      summary: Node stops failing for {{ $labels.name }}
      description: >
        PowerfulSeal has been unable to kill node {{ $labels.name }} with UID
        {{ $labels.uid }} for a while now.
    labels:
     type: mute
  - alert: command_executions_failing
    expr: rate(seal_execute_failed_total{status="failed"}[5m]) > 0
    for: 5m
    annotations:
      summary: Node command executions failing for {{ $labels.name }}
      description: >
        PowerfulSeal has been unable to successfully execute commands on node
        {{ $labels.name }} with UID {{ $labels.uid }} for a while now.
    labels:
     type: seal
  - alert: filtering_to_empty_set
    expr: rate(seal_empty_filter_total[10m]) > 0
    for: 10m
    annotations:
      summary: Filtering stage returning empty sets
      description: >
        The filtering stage is returning empty sets.
    labels:
     type: seal
  - alert: matching_to_empty_set
    expr: rate(seal_empty_match_total[10m]) > 0
    for: 10m
    annotations:
      summary: Matching stage returning empty sets
      description: >
        The matching for {{ $labels.source }} is returning empty sets
    labels:
     type: seal
```

`/etc/alertmanager.yml`:
```yaml
global:
  resolve_timeout: 5m

route:
  receiver: 'default'
  routes:
  - receiver: 'seal'
    match:
      type: seal

# A list of notification receivers.
receivers:
  - name: default
    email_configs:

  - name: seal
    email_configs:

```

### CODE REVIEW

**Comment by seeker89** [comment]
This looks nice and solid. Good work !

I like the approach and I like the implementation. A couple of housekeeping items, and we should be good to go:

- [ ] merge the master into this branch and resolve the conflicts, so that we get the `py27` running in Travis,
- [ ] similar to the previous PR with `py27` support, I think it deserves a separate page of docs, with at the minimum:
  - the info from this PR,
  - checked in files for the grafana dashboard, alert manager etc,
  - step by step guide of how to get all of that installed in your prometheus setup
- [ ] bump up the version to 1.4.0

Thanks so much for this nice contribution !

**Comment by kz** [comment]
Thanks for the review! I have made the requested contributions. Please let me know what you think.

File: setup.py [hunk 1]
```diff
@@ -25,6 +25,8 @@ def read(fname):
         'jsonschema>=2.6.0,<3',
         'boto3>=1.5.15,<2.0.0',
         'future>=0.16.0,<1',
+        'boto3>=1.5.15,<2.0.0',
```
Comment (seeker89) [inline setup.py]: This seems to be a duplicate line.

File: METRICS.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
+# Metric Collection
```
[the remark this answers is not in the document]
Comment (kz) [reply METRICS.md]: Addressed!

**Comment by kz** [comment]
Thanks for the review! The cosmetic amendments have been made.

**Pull request merged**